packages/server/src/unstable-core-do-not-import/stream/jsonl.ts

```diff
@@ -266,20 +277,62 @@ function createBatchStreamProducer(opts: ProducerOptions) {
   for (const [key, item] of Object.entries(data)) {
     newHead[key] = encode(item, [key]);
   }
-  pending.activate();
 
-  return [newHead, stream.readable] as const;
+  yield newHead;
+
+  // Process all async iterables concurrently by racing their next values
+  try {
+    // Track pending promises for the next value of each iterator
+    const nextPromises = new Map<
+      ChunkIndex,
+      Promise<IteratorResult<ChunkData, ChunkData>>
+    >();
+
+    // Continue processing while there are still active iterators
+    while (iterators.size > 0) {
+      // Ensure we have a pending promise for each iterator's next value
+      for (const [idx, it] of iterators.entries()) {
+        if (!nextPromises.has(idx)) {
+          nextPromises.set(idx, it.next());
+        }
+      }
```

Comment:
this could probably be done a bit more optimally

it's a bit of an unecessary loop in a loop

feel free to have a play @oleasteo :)